Throw an Exception instead of an error on Cache::write() fail. See original #16932

fr3nch13 opened the pull request:
Closing #16932 and opening this one so the branches line up.

fr3nch13 committed:
Throw an Exception instead of a error on write fail.

Throw a CacheWriteException when a Cache::write() fails, instead of triggering an error.
This allows the developer more control over how to handle a Cache::write() issue.

fr3nch13 committed:
Update CacheTest.php

fr3nch13 committed:
Update ci.yml

Moving the cs-stan job, and requiring the testsuites to have cs-stan be successful first.
This should catch cs issues before using other resources.

fr3nch13 committed:
Revert "Update ci.yml"

This reverts commit 9827f8b7a5a7db49f279e6e4887cd752209c489c.

fr3nch13 committed:
Moving Cache Exceptions to their own namespace.

fr3nch13 committed:
Oops. Fixed CS errors.

fr3nch13 committed:
Adding class alias as per Mark's recommendation

fr3nch13 committed:
Update InvalidArgumentException.php

fr3nch13 committed:
Merge branch 'cakephp:4.x' into 4.x

fr3nch13 committed:
Adding test for class alias.

fr3nch13 committed:
Update CacheTest.php

Try catching the aliased exception as would happen in apps.

fr3nch13 committed:
Try how class_alias is being used elsewhere here.

fr3nch13 committed:
Update InvalidArgumentException.php

fr3nch13 committed:
Update InvalidArgumentException.php

fr3nch13 committed:
Update CacheTest.php

fr3nch13 committed:
Update CacheTest.php

fr3nch13 committed:
Update CacheTest.php

Maybe phpcs will accept this? I really don't want to add to the neon file for this.

fr3nch13 committed:
Update CacheTest.php

fr3nch13 committed:
Meta update.

fr3nch13 committed:
Merge branch 'cakephp:4.x' into 4.x

fr3nch13 committed:
Adding stub file.

fr3nch13 committed:
Merge branch '4.next' into 4.x

fr3nch13 committed:
Merge branch 'cakephp:4.next' into 4.next

fr3nch13 committed:
Update StringTemplateTest.php

fr3nch13 committed:
Added test for aliased Exception.

Added this here to hopefully cover codecov's coverage report.
I'm not sure if this is the cake way, but I based it off of how Console\Command is tested.

fr3nch13 committed:
Maybe phpstan will like this better?

fr3nch13 commented:
I guess my phpstan/phpunit coverage isn't as good as I thought. I'm not sure what needs to be done so that phpstan will pass, and codecov will pas as well. Anyone have any ideas/pointers?

jamisonbryant commented:
For the stan fail, could this be the issue?

```
$this->expectException('Cake\Cache\InvalidArgumentException');
```

It looks like the path is supposed to be `Cake\Cache\Exception\InvalidArgumentException`, no?

othercorey commented on tests/TestCase/Cache/InvalidArgumentExceptionTest.php:
```diff
@@ -0,0 +1,36 @@
+     */
+    public function testThrowingOldException(): void
+    {
+        $this->expectException(\Cake\Cache\InvalidArgumentException::class);
```
Use a `use` statement instead of the full path.

dereuromark replied:
Shouldn't cs fix this?

othercorey replied:
I think it is failing, just trying to be helpful.

dereuromark replied:
No, thing is CI is green on that part
We might need a sniffer rule here

ADmad replied:
The sniff for this exists in the our v5 ruleset, not v4.

fr3nch13 replied:
It was passing, but codecov was failing as it was lowering coverage, that's why I added it. I can remove it.

fr3nch13 replied:
I removed it, and codecov dropped it's coverage percentage. Will that sniff in 5.x cover it?

fr3nch13 committed:
Delete InvalidArgumentExceptionTest.php

ADmad commented on tests/TestCase/View/StringTemplateTest.php:
```diff
@@ -80,6 +80,27 @@ public function testAddWithInvalidTemplate(): void
+            'invalid' => false,
+        ];
+        $this->expectException(InvalidArgumentException::class);
+        $this->template->add($templates);
```
How are these tests related to your changes?

fr3nch13 replied:
It was related to the other pull request. It checks for more then just a null template. I removed it from here.

fr3nch13 committed:
Update StringTemplateTest.php

It was a layover from previous commits.
It's supposed to check for other types other than just null. I removed it.

ADmad commented on src/Cache/Exception/CacheWriteException.php:
```diff
@@ -0,0 +1,27 @@
+ *
+ * @copyright     Copyright (c) Cake Software Foundation, Inc. (https://cakefoundation.org)
+ * @link          https://cakephp.org CakePHP(tm) Project
+ * @since         4.4.12
```
```suggestion
 * @since         4.5.0
```

fr3nch13 replied:
Should the sniff be brought over as well for the exception?

fr3nch13 committed:
Update src/Cache/Exception/CacheWriteException.php

Co-authored-by: ADmad <ADmad@users.noreply.github.com>

fr3nch13 commented:
Is there anything else you all need for this pull?

fr3nch13 commented:
So this should be passing with the exception of codecov/patch. What's needed to make it happy?

othercorey commented on tests/TestCase/Cache/Engine/MemcachedEngineTest.php:
```diff
@@ -18,9 +18,9 @@
+use Cake\Cache\Exception\InvalidArgumentException;
 use Cake\TestSuite\TestCase;
 use DateInterval;
-use InvalidArgumentException;
```
Looks like this test is using \InvalidArgumentException and shouldn't be changed.

othercorey commented on VERSION.txt:
```diff
@@ -16,4 +16,4 @@
 // +--------------------------------------------------------------------------------------------+ //
 ////////////////////////////////////////////////////////////////////////////////////////////////////
-4.5.0-dev
+4.5.0-dev
```
Can you revert this file?
